## This notebook is for practicing techniques for sparse matrices in python.

In [76]:
import numpy as np
from numpy import *
import scipy
from scipy import sparse
from scipy.stats import uniform
from scipy.sparse import csr_matrix, random

### 1. A sparse matrix in two formats
COO and CSC formats: both use 3 1d arrays, which store: 
* The row indices of non-zero elements. 
* the column indices of non-zero elements.   
* the values of the non-zero elements.  
They differ in the order in which the info is stored, which will become apparent when I print them out.   
CSC = compressed sparse column  
CSR = compressed sparse row  
COO = coordinate list 

In [10]:
row_ind = np.array([0, 1, 1, 3, 5]) # store rows
col_ind = np.array([0, 2, 4, 3, 4]) # store columns
vals = np.array([1,2,3,4, 5], dtype=int) # store data as ints

In [11]:
mat_coo = sparse.coo_matrix((vals, (row_ind, col_ind))) # calls the sparse matrix creating function in scipy
print (mat_coo) # print out the coords of the sparse matrix next to each value
# this prints it out in the order of lowest row index first

  (0, 0)	1
  (1, 2)	2
  (1, 4)	3
  (3, 3)	4
  (5, 4)	5


In [12]:
print(mat_coo.toarray()) #prints the sparse matrix out as a 2d regular matrix

[[1 0 0 0 0]
 [0 0 2 0 3]
 [0 0 0 0 0]
 [0 0 0 4 0]
 [0 0 0 0 0]
 [0 0 0 0 5]]


In [14]:
print(mat_coo.tocsc()) # print out the matrix in CSC format, which goes order to lowest col index first

  (0, 0)	1
  (1, 2)	2
  (3, 3)	4
  (1, 4)	3
  (5, 4)	5


In [33]:
print(mat_coo.tocsr()) # print out the matrix in CSR format, which goes lowest row index first

  (0, 0)	1
  (1, 2)	2
  (1, 4)	3
  (3, 3)	4
  (5, 4)	5


### 2. Making a sparse matrix out of a dense matrix  and comparing how much space a sparse matrix takes up in memory when it is stored as a regular ol'matrix versus using sparse matrix data structures. 

In [58]:
np.random.seed(seed=15) # going to make a 2d 1,000x1,000 matrix out of random numbers
data = uniform.rvs(size = 1000000, loc = 0, scale = 2) #  1x1,000,000 array
data = np.reshape(data, (1000,1000)) # make it square

In [59]:
data # print it out

array([[1.69763539, 0.35779185, 0.10872643, ..., 1.16436342, 1.90517905,
        1.88433106],
       [0.13825406, 1.77456479, 1.91159446, ..., 0.30748295, 1.4311104 ,
        0.58903344],
       [1.49623278, 0.73519249, 0.38666372, ..., 1.65577582, 1.37665664,
        1.11256492],
       ...,
       [0.01241248, 0.97766803, 0.24634209, ..., 0.03788967, 1.63902288,
        1.86843639],
       [1.80668531, 1.92914304, 1.05308499, ..., 1.80200817, 0.73661862,
        0.06122685],
       [1.21469923, 1.51414501, 0.44636327, ..., 0.00653099, 1.87058806,
        1.30906423]])

In [60]:
tol = 1
data[data < tol] = 0 # set all values < tol to zero. 

In [61]:
data # print out data now that it has been made sparse

array([[1.69763539, 0.        , 0.        , ..., 1.16436342, 1.90517905,
        1.88433106],
       [0.        , 1.77456479, 1.91159446, ..., 0.        , 1.4311104 ,
        0.        ],
       [1.49623278, 0.        , 0.        , ..., 1.65577582, 1.37665664,
        1.11256492],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 1.63902288,
        1.86843639],
       [1.80668531, 1.92914304, 1.05308499, ..., 1.80200817, 0.        ,
        0.        ],
       [1.21469923, 1.51414501, 0.        , ..., 0.        , 1.87058806,
        1.30906423]])

In [62]:
data_csr = sparse.csr_matrix(data) # make the newly-sparse matrix data into a csr 
data_csr

<1000x1000 sparse matrix of type '<class 'numpy.float64'>'
	with 500379 stored elements in Compressed Sparse Row format>

In [63]:
print(data_csr)

  (0, 0)	1.6976353945371574
  (0, 5)	1.0600004497908506
  (0, 10)	1.835259795673125
  (0, 12)	1.4355473749471122
  (0, 13)	1.731430068380308
  (0, 14)	1.6141589644907515
  (0, 20)	1.997086805454344
  (0, 22)	1.5210205478578993
  (0, 24)	1.0194306116353093
  (0, 25)	1.8900768343808854
  (0, 29)	1.0766975243138512
  (0, 31)	1.075490438837672
  (0, 32)	1.3312550756497088
  (0, 34)	1.2460372386691208
  (0, 35)	1.2854497704904226
  (0, 43)	1.3913556212782552
  (0, 48)	1.2789788566277585
  (0, 50)	1.9492132714390935
  (0, 53)	1.368092424626819
  (0, 55)	1.5420722266785365
  (0, 56)	1.5835990027389495
  (0, 62)	1.4550450971558246
  (0, 64)	1.040992949619482
  (0, 65)	1.3691380858070283
  (0, 67)	1.2069461146225098
  :	:
  (999, 949)	1.5261524403354283
  (999, 951)	1.5003886201776993
  (999, 952)	1.8044932660284312
  (999, 955)	1.670238263715502
  (999, 962)	1.779174533364403
  (999, 964)	1.9082466438679633
  (999, 966)	1.1417847973094921
  (999, 967)	1.7870182833742627
  (999, 968)	1.95309596

In [64]:
datacsc = sparse.csc_matrix(data) # make a csc version of data
datacsc

<1000x1000 sparse matrix of type '<class 'numpy.float64'>'
	with 500379 stored elements in Compressed Sparse Column format>

In [65]:
print(datacsc)

  (0, 0)	1.6976353945371574
  (2, 0)	1.4962327777221502
  (3, 0)	1.221968611657076
  (6, 0)	1.1643138247818532
  (7, 0)	1.7245302047458626
  (10, 0)	1.3883659085864366
  (20, 0)	1.8932546668328816
  (22, 0)	1.5880483671062025
  (24, 0)	1.2016936212171978
  (26, 0)	1.250517913033949
  (27, 0)	1.1404558698953529
  (28, 0)	1.40170907805162
  (30, 0)	1.338757841954658
  (31, 0)	1.1763956536485651
  (32, 0)	1.0168136075883467
  (33, 0)	1.8767983775980537
  (35, 0)	1.2158084097563466
  (37, 0)	1.4545006027471097
  (38, 0)	1.4254488381113284
  (40, 0)	1.2835319083678796
  (42, 0)	1.2219457017152893
  (43, 0)	1.479767159147141
  (48, 0)	1.9830893220526231
  (49, 0)	1.2099444488965092
  (50, 0)	1.6781807387705583
  :	:
  (939, 999)	1.8069775865798317
  (940, 999)	1.6939901963983188
  (941, 999)	1.6860697952891723
  (942, 999)	1.5286775812577156
  (943, 999)	1.6095289648240778
  (947, 999)	1.6232696072939619
  (956, 999)	1.3226466426926344
  (958, 999)	1.371252758646771
  (959, 999)	1.2644142354

In [68]:
data_size = (data.nbytes/(1024**2))
data_size # print out the size of data matrix in kb

7.62939453125

In [69]:
datacsc_size = (datacsc.data.size/(1024**2))
datacsc_size # print out the size of data_csc version in kb

0.47719860076904297

Notes: Comparing the size difference for different sizes of the sparse matrix.  
For a 4x4 matrix: the non-sparse version is 0.125 Kb and the version that uses the sparse structures is 0.005 Kb.  
For a 1,000 x 1,000 matrix: the non-sparse version is 7.63 Mb and the version that uses the sparse structures is 0.477 Mb

### 3. Using scipy.sparse to quickly convert a sparse matrix back to a dense matrix

In [75]:
A = datacsc.todense()
A # print out matrix A (it's 1,000x1,000, won't show the entire thing)

matrix([[1.69763539, 0.        , 0.        , ..., 1.16436342, 1.90517905,
         1.88433106],
        [0.        , 1.77456479, 1.91159446, ..., 0.        , 1.4311104 ,
         0.        ],
        [1.49623278, 0.        , 0.        , ..., 1.65577582, 1.37665664,
         1.11256492],
        ...,
        [0.        , 0.        , 0.        , ..., 0.        , 1.63902288,
         1.86843639],
        [1.80668531, 1.92914304, 1.05308499, ..., 1.80200817, 0.        ,
         0.        ],
        [1.21469923, 1.51414501, 0.        , ..., 0.        , 1.87058806,
         1.30906423]])

### 4. Dictionary of Keys (DOK) format for sparse matrices  
This format stores the (row, column) addresses of non-zero matrix elements with the value of those matrix elements in a dictionary. A big advantage of this format is O(1) access to any one matrix element! That's fast :) 

In [77]:
np.random.seed(10)

In [82]:
B = random(100,100, format='dok', density=0.4)
B

<100x100 sparse matrix of type '<class 'numpy.float64'>'
	with 4000 stored elements in Dictionary Of Keys format>

In [83]:
print(B)

  (0, 0)	0.3749237899520337
  (5, 0)	0.791848875730182
  (7, 0)	0.4918822931518011
  (8, 0)	0.2193978681508716
  (16, 0)	0.39499386437644624
  (17, 0)	0.9979835437201027
  (21, 0)	0.26275992513143287
  (22, 0)	0.7905840428214183
  (24, 0)	0.5041066951871328
  (25, 0)	0.7898718627040401
  (28, 0)	0.8354354411284687
  (32, 0)	0.8213204971343881
  (38, 0)	0.6975273169557853
  (40, 0)	0.8970426093946828
  (42, 0)	0.3392807852974963
  (43, 0)	0.17739441122484745
  (46, 0)	0.7872084410190475
  (47, 0)	0.8719148929149805
  (48, 0)	0.6202480170303764
  (53, 0)	0.06405129462208681
  (56, 0)	0.44940250188552766
  (57, 0)	0.8709031210229158
  (60, 0)	0.668692030522325
  (63, 0)	0.08987706668995366
  (64, 0)	0.5799691350890243
  :	:
  (22, 99)	0.035492403023591024
  (24, 99)	0.6746728370182843
  (30, 99)	0.09941951659322468
  (33, 99)	0.8839744629880142
  (36, 99)	0.7698059738831209
  (37, 99)	0.979724455937658
  (39, 99)	0.6786502136123818
  (44, 99)	0.6412862101042118
  (52, 99)	0.10409372169860

In [84]:
print(B.size/(1024)) # print out size of DOK format matrix in Kb

3.90625


### 5. List of Lists (LIL) sparse matrix storage format  
Stores info as a 1xd array, where each row of the array is a list of the coordinates and the value. 

In [85]:
C = random(100,100, format='lil', density=0.4)
C

<100x100 sparse matrix of type '<class 'numpy.float64'>'
	with 4000 stored elements in List of Lists format>

In [86]:
C.toarray()

array([[0.59290371, 0.95854233, 0.40874144, ..., 0.        , 0.        ,
        0.39675487],
       [0.        , 0.        , 0.48688697, ..., 0.15591546, 0.        ,
        0.        ],
       [0.99374719, 0.21804664, 0.        , ..., 0.36729168, 0.64846572,
        0.44821838],
       ...,
       [0.        , 0.        , 0.58290522, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.54568535, 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.9456578 , ..., 0.        , 0.        ,
        0.        ]])

In [87]:
C.data

array([list([0.5929037058690344, 0.9585423299016501, 0.4087414445605745, 0.3333028700231948, 0.0548935968297104, 0.2317869306911171, 0.5229058149275524, 0.21473520253168543, 0.6442168028299703, 0.7332473631495394, 0.14828921500332082, 0.886504866628575, 0.3731788751225241, 0.341687228292491, 0.005723274012297663, 0.26604548164338804, 0.2124744845623644, 0.05579351483207784, 0.6528655549649394, 0.5290854296853541, 0.054414159396688366, 0.9374274938007158, 0.10743637857776756, 0.17404721939533663, 0.8968259755163772, 0.3265027803494479, 0.416457948033486, 0.1327539644970509, 0.4366762020278133, 0.94423858146629, 0.4400867000689589, 0.8943496238813904, 0.14244645680189172, 0.8422877257323399, 0.8197924101639856, 0.5079427563225818, 0.900819931918646, 0.17691904364937527, 0.3188179714081768, 0.6270579834863115, 0.5651702633160847, 0.2775843588985022, 0.650864378677482, 0.5884529886798554, 0.8531095753177109, 0.9156741142006313, 0.5106530062096243, 0.3535461503705518, 0.11481510905827341, 0

In [88]:
C.rows

array([list([0, 1, 2, 5, 8, 11, 15, 16, 17, 19, 20, 22, 23, 26, 28, 31, 33, 34, 35, 37, 39, 40, 48, 50, 51, 52, 54, 57, 58, 59, 60, 61, 62, 64, 66, 67, 69, 70, 72, 73, 76, 78, 80, 81, 83, 85, 87, 89, 91, 92, 99]),
       list([2, 4, 5, 7, 9, 11, 12, 15, 17, 19, 20, 22, 28, 29, 30, 34, 35, 36, 37, 38, 42, 43, 45, 47, 48, 53, 56, 57, 58, 60, 62, 63, 66, 73, 76, 79, 80, 86, 88, 91, 97]),
       list([0, 1, 5, 6, 7, 8, 10, 13, 17, 18, 22, 27, 29, 30, 31, 35, 36, 38, 39, 40, 41, 42, 49, 50, 51, 53, 55, 57, 58, 59, 60, 61, 65, 67, 70, 71, 76, 80, 81, 82, 83, 86, 87, 88, 91, 92, 95, 96, 97, 98, 99]),
       list([0, 1, 4, 8, 9, 11, 15, 16, 26, 33, 34, 35, 41, 42, 43, 44, 45, 47, 51, 52, 55, 56, 58, 59, 60, 61, 62, 64, 66, 68, 70, 74, 75, 80, 86, 87, 89, 90, 91, 97]),
       list([3, 4, 6, 8, 9, 14, 15, 17, 20, 22, 23, 32, 33, 35, 39, 40, 41, 43, 44, 45, 46, 47, 49, 51, 53, 54, 56, 58, 59, 62, 66, 70, 71, 72, 78, 89, 90, 92, 98, 99]),
       list([0, 2, 5, 7, 9, 11, 13, 14, 15, 17, 22, 27, 33,

In [89]:
print(C.size/1024) # size of C in Kb

3.90625


### References: 

List of tutorials followed or sites with helpful discussion read while making this. 
1. https://cmdlinetips.com/2018/03/sparse-matrices-in-python-with-scipy/   
2. Setting array elements = some value, subject to some condition, without bothering to write a for loop: https://stackoverflow.com/questions/28430904/set-numpy-array-elements-to-zero-if-they-are-above-a-specific-threshold  
3. Nice command from scipy.sparse: https://www.educative.io/edpresso/sparse-matrices-in-python  
4. DOK and LIL format: https://rushter.com/blog/scipy-sparse-matrices/  
5. DOK format: https://scipy-lectures.org/advanced/scipy_sparse/dok_matrix.html  